scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Record.java

```diff
@@ -1103,7 +1105,15 @@ private DTLSMessage handshakeMessageFromByteArray(byte[] decryptedMessage) throw
 			}
 			LOGGER.debug(msg.toString(), parameter);
 		}
-		return HandshakeMessage.fromByteArray(decryptedMessage, parameter, getPeerAddress());
+		HandshakeMessage handshakeMessage = HandshakeMessage.fromByteArray(decryptedMessage, parameter, getPeerAddress());
+		if (handshakeMessage.getMessageType() == HandshakeType.FINISHED && epoch == 0) {
+			LOGGER.debug("FINISH with epoch 0 from peer [{}]!",
+					peerAddress);
+			AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.UNEXPECTED_MESSAGE,
+					peerAddress);
+			throw new HandshakeException("FINISH with epoch 0!", alert);
+		}
+		return handshakeMessage;
```

Comment:
It sounds like Handshaker logic so I would move it in an `Handshaker` :thinking:

Reply:
My idea was, that a FINISH epoch 0 is a invalid record. But I can move it into the handshaker.